### ISSUE TITLE
RequireJs is not resolving for some js files

### ISSUE THREAD

**Issue by wellingtonamaralleitao** [requirejs/requirejs#1544]
I've been stuck on this for a long time.

I don't get why doesn't find my module on this path:

`16 05 2016 19:30:53.478:WARN [web-server]: 404:    /base/skins/webmail2016/assets/vendors/jquery.validate.js`

And thats making this error happen when I call validate on my login.js function:
```
Chrome 49.0.2623 (Linux 0.0.0) ERROR
  Uncaught TypeError: $(...).validate is not a function
  at /home/franklinsantos/webmail2016/skins/webmail2016/assets/js/login.js:21

```

My `test-main.js`

``` javascript
require.config({
  // Karma serves files under /base, which is the basePath from your config file
  baseUrl: '/base',

  paths: {
    'jquery': 'skins/webmail2016/assets/vendors/jquery.min',
    'jquery.validate.unobtrusive': 'skins/webmail2016/assets/vendors/jquery.validate.unobtrusive',
    'jasmine': 'skins/webmail2016/spec/javascripts/helpers/jasmine-jquery',
    'login': 'skins/webmail2016/assets/js/login'
  },
  shim: {
    'jquery.validate': ['jquery'],
    'jquery.validate.unobtrusive': ['jquery', 'jquery.validate']
  },
```

My `karma.config.js`

```
basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine-jquery', 'jasmine', 'requirejs'],


    // list of files / patterns to load in the browser
    files: [
      {pattern: 'skins/webmail2016/assets/**/*.js', included: false},
      {pattern: 'skins/webmail2016/spec/**/*[Sspec].js', included: false},
      {pattern: 'skins/webmail2016/spec/javascripts/fixtures/*.html', included: false, served: true}, // Fixtures files
      'test-main.js'
    ],

```

Here is where I define my modules to be used.

`login_spec.js`  

```
define(['login', 'jquery', 'jquery.validate', 'jquery.validate.unobtrusive'], function($) {
  describe('Suit tests for locamail2016 - login', function(){

    beforeEach(function(){
      jasmine.getFixtures().fixturesPath = 'base/skins/webmail2016/spec/javascripts/fixtures';
      loadFixtures('login_fix.html');
      window.locamail.login.init();
    });

    it('should toogle between hide and show password texts', function(){
      expect($('#showpass')).toHaveText('Exibir');
    });

  });

});

```

**Here is the workingtree**

![](https://snag.gy/YXaW1u.jpg)

**Comment by ScottFreeCode** [comment]
- Your test-main.js file's `paths` appears to be missing the entry for `jquery.validate`. Are you sure this is up to date and complete?
- Your test-main.js file uses `/base` as the base url, but karma.conf.js file appears to set the basePath to ''; what happens if you comment out that line in karma.conf.js?
- The directory tree isn't showing the contents of the vendor folder where jquery.validate.js is supposed to be, so from what I see here it's possible (however unlikely) that the file is missing.
- It is also possible that this is a problem with Karma's Require plugin, or a matter of configuring it where you'd get better advice from someone familiar with that project more than someone familiar with RequireJS in itself.

**Comment by wellingtonamaralleitao** [comment]
First of all thanks for you help...
 - I was editing the `test-main.js` so I forgot to put the reference to `jquery.validate.js` back before pasting to here, sorry. But thats not the problem.
 - Nothing happens...
- Sublime was not auto-refreshing the directory tree. But the file is there.

I'm not having 404 error anymore but still doesnt recognize my `jquery.validate` function on `login.js`

**Comment by ScottFreeCode** [comment]
Well, that's good; so the file not being found doesn't seem to be the issue after all... I see login_spec.js there, but my other hunch is about login.js (where the error is actually occurring); given that jquery.validate is using shim config, I'm wondering whether login.js is referring to it through a global variable or an AMD dependency/parameter.

**Comment by wellingtonamaralleitao** [comment]
I'm still new with modules in javascript. So I'm not really sure if I need to pass the reference into my js file as a parameter like I'm doing with jquery.

`login.js` looks like this...
``` javascript
var locamail = locamail || {};
locamail.login = (function($) {
  'use strict';

...

function validateForm() {
    $(  '#loginform'  ).validate({

...
}(jQuery));

```

**Comment by ScottFreeCode** [comment]
So, if I'm following this right, login.js is ending up loaded before JQuery (or at least before the validate plugins), hence the attempt to use `jquery(...).validate` failing. (The order scripts actually load with RequireJS and other AMD loaders is usually hard to predict except for the dependency tree, and even then the actual top-level code that calls `define` doesn't run in the same order that the callbacks end up getting called.) It's set up with an immediately-invoked function expression so its local variables are isolated from other scripts, but it's not actually set up as a module (just namespacing its output on a global variable), so RequireJS isn't being directed that JQuery etc. needs to be loaded first. A simple fix would be to add it to the shim config, which is [designed to add module-like behavior to these old-style scripts without changing them into modules](http://requirejs.org/docs/api.html#config-shim); the simple form you've already got for JQuery just gives RequireJS the list of dependencies to load first (which would be enough: `'login': ['jquery.validate'],`), but you can also configure it like this:

```js
shim: {
  'someModule': {
    deps: ['oneDependency', 'anotherDependency'],
    exports: 'the.global.variable.this.module.sets.up'
  }
},
```

...or in your case:

>in test-main.js

```js
shim: {
  'login': {
    deps: ['jquery.validate'],
    exports: 'locamail.login'
  },
  'jquery.validate': ['jquery'],
  'jquery.validate.unobtrusive': ['jquery', 'jquery.validate']
},
```

This would allow AMD modules (made with `define`) or top-level AMD-aware code using the callback-based `require` to get the `window.locamail.login` variable passed as a parameter to the callback (the factory in the case of `define`), as if it were an AMD module. That offers a little more flexibility: the modules using login.js would not need to specify the global variable, and login.js could be converted to an AMD module without needing to change the others. (I'll have more examples of these after a little more explanation.) If you'd prefer, though, just the same type of dependency list as you have for JQuery would do.

Speaking of using these with AMD modules, though: another thing is the login_spec.js's dependency list, which is currently not going to get anything for the `$` parameter. The factory callback function you pass to `define` should have parameters matching the libraries in the array passed to `define` before it; that is, the first library ("login" in your case) will, once it's loaded, be the first parameter to your function (the `$` you've got in there so far; but, without configuring shim exports, RequireJS has no way to tell which variable set by login.js it should use as the parameter, so it doesn't have anything to pass as that parameter), the second library ("jquery") will be the second parameter (which currently isn't specified in the callback), etc. (You can have more libraries than parameters, but this is usually good mainly for libraries that provide side effects when they're loaded at all.) It looks like you want to have the `$` parameter correspond to JQuery, so you'd want to either move JQuery to the start of the list:

>login_spec.js beginning change to keep using login.js as a non-module script

```js
define(['jquery', 'login', 'jquery.validate', 'jquery.validate.unobtrusive'], function($) {
```

...or add a parameter for the login.js module -- this goes along with the use of `exports` in `shim` for "login", described above:

>login_spec.js for use with AMD-ified login.js (whether by upgrading login.js or by shim `exports`)

```js
define(['login', 'jquery', 'jquery.validate', 'jquery.validate.unobtrusive'], function(login, $) {
  describe('Suit tests for locamail2016 - login', function(){

    beforeEach(function(){
      jasmine.getFixtures().fixturesPath = 'base/skins/webmail2016/spec/javascripts/fixtures';
      loadFixtures('login_fix.html');
      login.init();
    });

    it('should toogle between hide and show password texts', function(){
      expect($('#showpass')).toHaveText('Exibir');
    });

  });

});
```

To come back around to one last thing, if you wanted to upgrade "login" to an AMD module, you'd change the immediately-invoked function expression into a `define` call, list `jquery` and `jquery.validate` in its dependency list instead of in shim config and the immediate invoke of the function expression, and skip setting `window.locamail.login` -- what your function is returning, instead of being used to set `window.locamail.login`, will be provided by RequireJS (or any other AMD loader) as the parameter to other modules' `define` factories and to `require` callbacks. Something like this:

>login.js upgraded to AMD, for use with login_spec.js in the same style as if shim set `exports` for it

```js
define(['jquery', 'jquery.validate'], function($) {
  'use strict';

...
function validateForm() {
    $(  '#loginform'  ).validate({
...

return ...whatever you were returning before, just the same;
...
});
```

Nice and easy to upgrade since the original already uses a return to set its global variable.

There are more advanced options out there, typically by the name of "UMD" (Universal instead of Asynchronous) if you want to set up modules so that they don't require shim config but will work either (or both) with AMD and/or without it: https://github.com/umdjs/umd But if that's too complicated, I wouldn't worry about it; you've already got a couple choices here with whether to make login.js an AMD module or just add shim config for it and, if just shim config, whether to set it up so that the modules using it can still use it as if it were AMD.

**TLDR quick solution:** add `'login': ['jquery.validate'],` to the `shim` entry in your configuration so `jquery.validate` loads before login.js, and swap 'jquery' and 'login' in the list at the start of login_spec.js so that JQuery corresponds to the `$` parameter; but keep in mind there are more complete modularization options, and while they're fairly major changes conceptually they're pretty simple and small changes at the code level.
